cloudinit/net/freebsd.py

```diff
@@ -0,0 +1,164 @@
+from cloudinit import net
+from cloudinit import util
+from cloudinit.distros import rhel_util
+from cloudinit.distros.parsers.resolv_conf import ResolvConf
+
+from . import renderer
+
+LOG = logging.getLogger(__name__)
+
+
+class Renderer(renderer.Renderer):
+    resolv_conf_fn = 'etc/resolv.conf'
+    rc_conf_fn = 'etc/rc.conf'
+
+    def __init__(self, config=None):
+        if not config:
+            config = {}
+        self.dhcp_interfaces = []
+        self._postcmds = config.get('postcmds', True)
+
+    def _write_ifconfig_entries(self, settings, target=None):
+        ifname_by_mac = net.get_interfaces_by_mac()
+        for interface in settings.iter_interfaces():
+            device_name = interface.get("name")
+            device_mac = interface.get("mac_address")
+            if device_name and re.match(r'^lo\d+$', device_name):
+                continue
+            if device_mac and device_name:
+                cur_name = ifname_by_mac[device_mac]
+                if not cur_name:
+                    LOG.info('Cannot find any device with MAC %s', device_mac)
+                    continue
+                if cur_name != device_name:
+                    rhel_util.update_sysconfig_file(
+                        util.target_path(target, self.rc_conf_fn), {
+                            'ifconfig_%s_name' % cur_name: device_name})
+            elif device_mac:
+                device_name = ifname_by_mac[device_mac]
+
+            LOG.info('Configuring interface %s', device_name)
+            ifconfig = 'DHCP'  # default
+            for subnet in interface.get("subnets", []):
+
+                if subnet.get('type') == 'static':
+                    LOG.debug('Configuring dev %s with %s / %s', device_name,
+                              subnet.get('address'), subnet.get('netmask'))
+                # Configure an ipv4 address.
+                    ifconfig = (
+                            subnet.get('address') + ' netmask ' +
+                            subnet.get('netmask'))
+
+                # NOTE: Known limitation, we just set one subnet per interface
+                break
+
+            if ifconfig == 'DHCP':
+                self.dhcp_interfaces.append(device_name)
+            rhel_util.update_sysconfig_file(
+                util.target_path(target, self.rc_conf_fn), {
+                    'ifconfig_' + device_name: ifconfig})
+
+    def _write_route_entries(self, settings, target=None):
+        routes = list(settings.iter_routes())
+        for interface in settings.iter_interfaces():
+            subnets = interface.get("subnets", [])
+            for subnet in subnets:
+                if subnet.get('type') == 'static':
+                    gateway = subnet.get('gateway')
+                    if gateway:
+                        routes.append({
+                            'network': '0.0.0.0',
+                            'netmask': '0.0.0.0',
+                            'gateway': gateway})
```

Comment:
This looks strange to me.  Shouldn't the subnet provide at least a netmask?

Reply:
I suspect this is related to a default gateway?  But certainly we can have subnets that are not default gateways, so this is going to break.

Reply:
I was initially using a NoCloud+ENI configuration, and don't need that with OpenStack metadata format.